File: 2023/day_2/python/main.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


game_constraints = {
    "red": 12,
    "green": 13,
    "blue": 14,
}

class Day2:
    filename = "input.txt"
    file_content = []
    result = 0

    def read_file(self):
        with open(self.filename) as f:
            self.file_content = f.read().splitlines()

    def is_game_possible(self, input):
        i = input.find(":")
        game_id = int(input[5:i])

        game_results = input[i+2:]
        for result in game_results.split("; "):
            for color_result in result.split(", "):
                space_index = color_result.find(" ")
                num = int(color_result[:space_index])
                color = color_result[space_index+1:]
                if num > game_constraints[color]:
                    return game_id, False

        return game_id, True

    def get_game_power(self, input):
        i = input.find(":")
        lowest_possible = {
            "red": 0,
            "green": 0,
            "blue": 0,
        }

        game_results = input[i+2:]
        for result in game_results.split("; "):
            for color_result in result.split(", "):
                space_index = color_result.find(" ")
                num = int(color_result[:space_index])
                color = color_result[space_index+1:]
                if num > lowest_possible[color]:
                    lowest_possible[color] = num

        game_power = 1
        for color in lowest_possible.values():
            game_power *= color

        return game_power

    def task1(self):
        self.result = 0
        for game in self.file_content:
            game_id, result = self.is_game_possible(game)
            if result:
                self.result += game_id

    def task2(self):
        self.result = 0
        for game in self.file_content:
            self.result += self.get_game_power(game)


aoc = Day2()
aoc.read_file()
aoc.task1()
print("Result of task 1: {}".format(aoc.result))
aoc.task2()
print("Result of task 2: {}".format(aoc.result))

File: 2023/day_2/cpp/src/main.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <string>

std::map<std::string, int> game_constraints = {
    {"red", 12},
    {"green", 13},
    {"blue", 14},
};

std::pair<int, bool> isGamePossible(std::string game_line) {
    int i = game_line.find(":");
    int game_id = std::stoi(game_line.substr(5, i-5));
    std::string game_results = game_line.substr(i+2);
    size_t pos;

    do {
        pos = game_results.find("; ");
        std::string token = game_results.substr(0, pos);
        game_results.erase(0, pos + 2);

        size_t inner_pos;
        do {
            inner_pos = token.find(", ");
            std::string color_result = token.substr(0, inner_pos);
            token.erase(0, inner_pos + 2);

            int space_index = color_result.find(" ");
            int num = std::stoi(color_result.substr(0, space_index));
            std::string color = color_result.substr(space_index + 1);

            if(num > game_constraints[color]) {
                std::pair<int, bool> result{game_id, false};
                return result;
            }
        } while(inner_pos != std::string::npos);

    } while(pos != std::string::npos);

    std::pair<int, bool> result{game_id, true};
    return result;
}

void solveTask1(const std::string *filename) {
    std::ifstream input_stream;

    try {
        input_stream.open(*filename);
    }
    catch (std::ifstream::failure e) {
        std::cerr << "Exception opening a file\n";
    }

    std::string line;
    int result = 0;
    while(std::getline(input_stream, line)) {
        std::pair<int, bool> game_possibility = isGamePossible(line);
        if(game_possibility.second) {
            result += game_possibility.first;
        }
    }

    std::cout << "Result of task 1: " << result << std::endl;
}

int main() {
    const std::string filename = "../../input.txt";

    solveTask1(&filename);

    return 0;
}
